File: Paid_ChatGPT/sistemaDeTiposRevisado/tipos.c
#include <stdio.h>
#include <stdlib.h>

typedef enum {
    INT,
    FLOAT,
    CHAR,
    STRING
} Tipo;

typedef struct {
    Tipo tipo;
    union {
        int i;
        float f;
        char c;
        char *s;
    } valor;
} Elemento;

double convertirANumero(Elemento e) {
    switch (e.tipo) {
        case INT:
            return (double)e.valor.i;

        case FLOAT:
            return (double)e.valor.f;

        case CHAR:
            return (double)(e.valor.c - '0');

        case STRING:
            return atof(e.valor.s);

        default:
            return 0;
    }
}

double promedioMayoresAUmbral(Elemento arreglo[], int n, Elemento umbral) {
    double suma = 0.0;
    int cantidad = 0;

    double umbralNum = convertirANumero(umbral);

    for (int i = 0; i < n; i++) {
        double valor = convertirANumero(arreglo[i]);

        if (valor > umbralNum) {
            suma += valor;
            cantidad++;
        }
    }

    return (cantidad == 0) ? 0 : suma / cantidad;
}

int main() {
    Elemento arreglo[] = {
        {CHAR,   {.c = '1'}},
        {FLOAT,  {.f = 2.5}},
        {INT,    {.i = 30}},
        {STRING, {.s = "40"}},
        {FLOAT,  {.f = 50.0}}
    };

    Elemento umbral = {INT, {.i = 25}};

    double promedio = promedioMayoresAUmbral(arreglo, 5, umbral);

    printf("Promedio: %.2f\n", promedio);

    return 0;
}

File: Paid_ChatGPT/sistemaDeTiposRevisado/tipos.rs
enum Valor {
    Int(i32),
    Float(f64),
    Char(char),
    Str(String),
}

fn convertir_a_numero(valor: &Valor) -> f64 {
    match valor {
        Valor::Int(i) => *i as f64,
        Valor::Float(f) => *f,
        Valor::Char(c) => c.to_digit(10).unwrap() as f64,
        Valor::Str(s) => s.parse::<f64>().unwrap(),
    }
}

fn promedio_mayores_a_umbral(arreglo: &[Valor], umbral: &Valor) -> f64 {
    let umbral_num = convertir_a_numero(umbral);

    let mut suma = 0.0;
    let mut cantidad = 0;

    for elemento in arreglo {
        let valor = convertir_a_numero(elemento);

        if valor > umbral_num {
            suma += valor;
            cantidad += 1;
        }
    }

    if cantidad == 0 {
        0.0
    } else {
        suma / cantidad as f64
    }
}

fn main() {
    let valores = vec![
        Valor::Char('1'),
        Valor::Float(2.5),
        Valor::Int(30),
        Valor::Str("40".to_string()),
        Valor::Str("50".to_string()),
    ];

    let umbral = Valor::Int(25);

    let promedio = promedio_mayores_a_umbral(&valores, &umbral);

    println!("{:.2}", promedio);
}